Add PDF protection blog post (#528)
metanorma/metanorma.org: _posts/2021-12-04-pdf-protection.adoc

---
layout: post
title: "Metanorma now offers PDF protection mechanisms!"
date: 2021-12-04
categories: documentation
authors:
  -
    name: Ronald Tse
    email: ronald.tse@ribose.com
    social_links:
      - https://github.com/ronaldtse
  -
    name: Alexander Dyuzhev
    email: dyuzhev@gmail.com
    social_links:
      - https://www.linkedin.com/in/alexander-dyuzhev/
      - https://github.com/Intelligent2013/

excerpt: >-
  Metanorma-generated PDFs now support encryption and other PDF protection
  features.
---
== Introduction

> There are times when a Metanorma document needs to be protected, such as:

Reviewer: ```suggestion
There are times when a Metanorma PDF document needs to be protected, such as:
```

* when sharing confidential information;

> * to prevent simplistic copy and pasting of content;

Reviewer: ```suggestion
* to prevent direct copy and pasting of content;
```

* to prevent printing the document; or
* to prevent high quality print reproduction of a document.

https://www.iso.org/standard/51502.html[ISO 32000-1:2008]
(free copy at https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf[Adobe])
defines PDF 1.7 which specifies several PDF protection mechanisms.

> As several SDOs that use Metanorma requested such functionality, we are happy to

Reviewer: ```suggestion
As several SDOs that use Metanorma have requested such functionality, we are happy to
```

announce that the following Metanorma PDF protection features are now generally
available:

* Require password to open document (PDF whole-document encryption);
* Disallow copying text or graphics;
* Disallow inserting, deleting or rotating pages;
* Disallow adding annotations or signatures;
* Disallow printing;
* Disallow high quality printing.

WARNING: Not all PDF protection features are implemented in PDF viewers! See below.

== PDF whole document protection

=== Require password to open document

Metanorma implements the PDF encryption mechanism specified in ISO 32000-1, 7.6,
originally from PDF 1.1.

This is the most popular protection measure where the entire PDF is
"`protected`", the reader cannot open the PDF file without the password given.

Typically, the PDF viewer will prompt the user to enter the password on loading
the PDF file. If the password entered is correct, the PDF loads. Otherwise,
the PDF load fails.

PDF encryption nowadays often relies on the AES algorithm, with either 256-bit
or 128-bit security.

While the PDF standard originally allows for both RC4 and AES encryption

> algorithms, Metanorma only supports AES, as RC4 has been shown have a number of

Reviewer: ```suggestion
algorithms, Metanorma only supports AES, as RC4 has been shown to have a number of
```

weaknesses (https://en.wikipedia.org/wiki/RC4[Wikipedia]).

Metanorma also defaults to using 256-bit for superior security. But you may find
legacy PDF readers only supporting 128-bit due to US export restrictions
previously in place.

In order to generate a encrypted Metanorma PDF output, you will need to provide
the following document attributes:

* `:pdf-encrypt: true` to specify encryption of PDF output

* `:pdf-encryption-length: {128,256}` to specify encryption strength
  (128-bit or 256-bit AES keys)

* `:pdf-user-password: {password}` to specify the "user password" needed to
  open the encrypted PDF document.

[example]
====
[source,adoc]
----
= My Confidential Document
:pdf-encrypt: true
:pdf-encryption-length: 256
:pdf-user-password: C4nt0p3nW1th0utTh1s
...
----
====

== PDF permissions

=== General and the "Owner Password"

PDF permissions are actions normally allowed with unprotected PDFs, but these

> actions can be selectively restricted by the author with a "`owner password`".

Reviewer: ```suggestion
actions can be selectively restricted by the author with an "`owner password`".
```

Author: Fixed in e11167a.

If the reader enters the designated "`owner password`", these restrictions can
be bypassed.

The owner password is set by the following document attribute:

* `:pdf-owner-password: {password}` specifies the PDF "owner password".

[example]
====
[source,adoc]
----
= Copy Pasta
:pdf-allow-copy-content: false
:pdf-owner-password: D0nt700k4tM3
...
----
====

=== Disallow copying text or graphics

A PDF file that is marked for copy production means its content is not
to be copied from the reader view. Commercial publishers typically apply this

> for copyright and content protection concerns (other than DRM).

Reviewer: ```suggestion
out of concerns for copyright and content protection (other than DRM).
```

> Sort of as a "`your eyes only`" document, where you can read but not copy out

Reviewer: ```suggestion
This makes the PDF generated something of a "`your eyes only`" document, where you can read but not copy out
```

content.

This permission can be disallowed using the following document attribute:

* `:pdf-allow-copy-content: false` specifies that copying of content is
disallowed.

[example]
====
[source,adoc]
----
= Copy Pasta
:pdf-allow-copy-content: false
:pdf-owner-password: D0nt700k4tM3
...
----
====
//:pdf-allow-edit-content: false

=== Disallow editing of content

Editing of PDF content can be disallowed to ensure that the document content remains as when it was originally published.

> This permission can be disallowed using the following document attribute:

Reviewer: ```suggestion
Editing can be disallowed using the following document attribute:
```

> * `:pdf-allow-assemble-document: false` specifies that editing of PDF content is disallowed.

Reviewer: ```suggestion
* `:pdf-allow-edit-content: false` specifies that editing of PDF content is disallowed.
```

Author: @opoudjis "edit content" and "assemble document" are different permissions.

[example]
====
[source,adoc]
----
= No Edit
:pdf-allow-edit-content: false
:pdf-owner-password: D0nt700k4tM3
...
----
====

=== Disallow inserting, deleting or rotating pages

A PDF file that is marked to disallow page modification means an unauthorized
user cannot modify its content by inserting, deleting or rotating pages.

This permission can be disallowed using the following document attribute:

> * `:pdf-allow-assemble-document: false` specifies that copying of content is

Reviewer: I don't like this, is "assemble document" an established term? Can we say "reassemble", "rearrange", or "interleave"?

Author: From ISO 32000-1:2008: "Assemble" is the correct term.

![image](https://user-images.githubusercontent.com/11865/145185403-0bbf0b8d-2f5e-428f-aeed-61f10c71bbd4.png)

This permission can be disallowed using the following document attribute:

* `:pdf-allow-edit-annotations: false` specifies that annotations and signatures cannot be added.

[example]
====
[source,adoc]
----
= Copy Pasta
:pdf-allow-edit-annotations: false
:pdf-owner-password: D0nt700k4tM3
...
----
====

=== Disallow printing

A PDF file marked "`not for printing`" means exactly that, the PDF content
cannot be printed on a printer.

Compliant PDF readers will block not only printing directly to a physical
printer, they would also block printing to files (e.g. features like "`print to
PDF`" or "`save as image`").

This permission can be disallowed using the following document attribute:

* `:pdf-allow-print: false` specifies that the PDF document cannot be printed.

[example]
====
[source,adoc]
----
= Print Me, Will You?
:pdf-allow-print: false
:pdf-owner-password: D0nt700k4tM3
...
----
====

=== PDF protection against high quality printing

For these PDFs, low quality prints are allowed, but not high resolution prints.
The intent for this limitation is to prevent an accurate physical reproduction
of documents that could potentially mislead someone that a print out (from this

> file) was authentic.

Reviewer: "authentic" is not the word you want here. Authorised? Official?

Author: Maybe? I can go with "authorised", but with "authentic" I mean whether the content has been maliciously adjusted.

The resulting print out is supposed to look like the original PDF but blurred
and of poor resolution.

> This permission can be disallowed using the following document attribute:

Reviewer: ```suggestion
This feature can be enabled using the following document attribute:
```

> * `:pdf-allow-print-hq: true` specifies that the PDF document can only be printed in low resolution.

Reviewer: ```suggestion
* `:pdf-allow-print-hq: false` specifies that the PDF document can only be printed in low resolution.
```

[example]
====
[source,adoc]
----
= Visual Transparency

> :pdf-allow-print-hq: true

Reviewer: ```suggestion
:pdf-allow-print-hq: false
```

[example]
====
[source,adoc]
----
= Visual Transparency
:pdf-allow-fill-in-forms: false
:pdf-allow-access-content: false
:pdf-encrypt-metadata: true
...
----
====

Non-standard PDF measures like DRM, which are popular with digital publishers
are also not described here.

=== Preparing a fully "locked-down" document

The following example is a good profile to create a fully locked-down document
(encrypted or not).

[example]
====
[source,adoc]
----
= Very Well Protected Document
:pdf-owner-password: 0wnerP4ss
:pdf-user-password: D0cR04d0n1y
:pdf-allow-print: false
:pdf-allow-copy-content: false
:pdf-allow-edit-content: false
:pdf-allow-edit-annotations: false
:pdf-allow-fill-in-forms: false
:pdf-allow-access-content: false
:pdf-allow-assemble-document: false
:pdf-allow-print-hq: false
:pdf-allow-fill-in-forms: false
:pdf-allow-access-content: false
:pdf-encrypt-metadata: true
----
====

== Compatibility concerns and caveats

=== Potential caveats

The aforementioned features come with a caveat -- not all PDF viewers today
fully support these standard PDF features.

A sobering fact: out of the implemented PDF protection features, only

> whole-document encryption works across all the popular PDF viewers we tested.

Reviewer: ```suggestion
whole-document encryption works across all the popular PDF viewers we have tested.
```

| Foxit PDF Reader  | 'Print' inactive, cannot be printed, does not prompt password from user | ✅
| Preview (macOS)   | Prints properly | ❌
| Skim (macOS)      | Prints properly | ❌
| Firefox (browser) | 'Print' is active and can be printed, moreover user can choose 'Save to PDF' and resulted PDF will be re-saved without any security features (Firefox uses an internal PDF printer) | ❌
| Safari (browser)  | Prints properly | ❌
| Microsoft Word    | Opens PDF file and converts into Microsoft Word DOCX | N/A

3+| *Option*: Disallow high-quality printing
| Adobe Reader      | Prints properly, resulting print out with reduced resolution | ✅
| Foxit PDF Reader  | Prints properly, in high quality | ❌
| Preview (macOS)   | Prints properly, in high quality | ❌
| Skim (macOS)      | Prints properly, in high quality | ❌
| Firefox (browser) | Prints properly, in high quality | ❌
| Safari (browser)  | Prints properly, in high quality | ❌
| Microsoft Word    | Opens PDF file and converts into Microsoft Word DOCX | N/A

|===

This means:

* The only protection measure that works across the board is
  "`whole PDF encryption`".

> * Adobe Reader implements all PDF protection mechanisms properly, but not the
>   other popular viewers.

Reviewer: ```suggestion
* Adobe Reader implements all PDF protection mechanisms properly, but the
  other popular viewers do not.
```
